components/script/dom/htmlinputelement.rs

```diff
@@ -63,16 +63,118 @@ const PASSWORD_REPLACEMENT_CHAR: char = '●';
+    fn is_textual(&self) -> bool {
+        match *self {
+            InputType::Color | InputType::Date | InputType::Datetime
+            | InputType::DatetimeLocal | InputType::Email | InputType::Hidden
+            | InputType::Month | InputType::Number | InputType::Range
+            | InputType::Search | InputType::Tel | InputType::Text
+            | InputType::Time | InputType::Url | InputType::Week => {
+                true
+            }
+
+            _ => false
+        }
+    }
+
+    fn is_textual_or_password(&self) -> bool {
+        self.is_textual() || *self == InputType::Password
+    }
+
+    fn to_str(&self) -> &str {
+        match *self {
+            InputType::Button => "button",
+            InputType::Checkbox => "checkbox",
+            InputType::Color => "color",
+            InputType::Date => "date",
+            InputType::Datetime => "datetime",
+            InputType::DatetimeLocal => "datetime-local",
+            InputType::Email => "email",
+            InputType::File => "file",
+            InputType::Hidden => "hidden",
+            InputType::Image => "image",
+            InputType::Month => "month",
+            InputType::Number => "number",
+            InputType::Password => "password",
+            InputType::Radio => "radio",
+            InputType::Range => "range",
+            InputType::Reset => "reset",
+            InputType::Search => "search",
+            InputType::Submit => "submit",
+            InputType::Tel => "tel",
+            InputType::Text => "text",
+            InputType::Time => "time",
+            InputType::Url => "url",
+            InputType::Week => "week",
+        }
+    }
+}
+
+impl<'a> From<&'a Atom> for InputType {
+    fn from(value: &Atom) -> InputType {
+        match value {
+            &atom!("button") => InputType::Button,
+            &atom!("checkbox") => InputType::Checkbox,
+            &atom!("color") => InputType::Color,
+            &atom!("date") => InputType::Date,
+            &atom!("datetime") => InputType::Datetime,
+            &atom!("datetime-local") => InputType::DatetimeLocal,
+            &atom!("email") => InputType::Email,
+            &atom!("file") => InputType::File,
+            &atom!("hidden") => InputType::Hidden,
+            &atom!("image") => InputType::Image,
+            &atom!("month") => InputType::Month,
+            &atom!("number") => InputType::Number,
+            &atom!("password") => InputType::Password,
+            &atom!("radio") => InputType::Radio,
+            &atom!("reset") => InputType::Reset,
+            &atom!("search") => InputType::Search,
+            &atom!("submit") => InputType::Submit,
+            &atom!("tel") => InputType::Tel,
+            &atom!("text") => InputType::Text,
+            &atom!("time") => InputType::Time,
+            &atom!("url") => InputType::Url,
+            &atom!("week") => InputType::Week,
+            v if *v == Atom::from("range") => InputType::Range,
```

Comment:
I couldn't get it to compile without this hack. The compiler said that `"range"` didn't match the `atom!` macro. I'm not sure what the right thing to do here is, but I suspect this isn't it! Please let me know.

Reply:
You'll need to add the "range" string to [this file](https://github.com/servo/servo/blob/master/components/atoms/static_atoms.txt) in order for the `atom!` macro to work with it.

Reply:
Thanks. Seems like the way atoms work could really use some documentation, I've had a lot of questions about this area of the code.